common/client/transaction_sender.go

```diff
@@ -0,0 +1,278 @@
+func (txSender *TransactionSender[TX, CHAIN_ID, RPC]) SendTransaction(ctx context.Context, tx TX) (SendTxReturnCode, error) {
+	txResults := make(chan sendTxResult)
+	txResultsToReport := make(chan sendTxResult)
+	primaryNodeWg := sync.WaitGroup{}
+
+	ctx, cancel := txSender.chStop.Ctx(ctx)
+	defer cancel()
+
+	healthyNodesNum := 0
+	err := txSender.multiNode.DoAll(ctx, func(ctx context.Context, rpc RPC, isSendOnly bool) {
+		if isSendOnly {
+			txSender.wg.Add(1)
+			go func() {
+				defer txSender.wg.Done()
+				// Send-only nodes' results are ignored as they tend to return false-positive responses.
+				// Broadcast to them is necessary to speed up the propagation of TX in the network.
+				_ = txSender.broadcastTxAsync(ctx, rpc, tx)
+			}()
+			return
+		}
+
+		// Primary Nodes
+		healthyNodesNum++
+		primaryNodeWg.Add(1)
+		go func() {
+			defer primaryNodeWg.Done()
+			result := txSender.broadcastTxAsync(ctx, rpc, tx)
+			select {
+			case <-ctx.Done():
+				return
+			case txResults <- result:
+			}
+
+			select {
+			case <-ctx.Done():
+				return
+			case txResultsToReport <- result:
+			}
+		}()
+	})
+
+	// This needs to be done in parallel so the reporting knows when it's done (when the channel is closed)
+	txSender.wg.Add(1)
+	go func() {
+		defer txSender.wg.Done()
+		primaryNodeWg.Wait()
+		close(txResultsToReport)
+		close(txResults)
+	}()
+
+	if err != nil {
+		txSender.wg.Wait()
+		return 0, err
+	}
+
+	txSender.wg.Add(1)
+	go txSender.reportSendTxAnomalies(tx, txResultsToReport)
+
+	return txSender.collectTxResults(ctx, tx, healthyNodesNum, txResults)
+}
+
+func (txSender *TransactionSender[TX, CHAIN_ID, RPC]) broadcastTxAsync(ctx context.Context, rpc RPC, tx TX) sendTxResult {
+	txErr := rpc.SendTransaction(ctx, tx)
+	txSender.lggr.Debugw("Node sent transaction", "tx", tx, "err", txErr)
+	resultCode := txSender.txErrorClassifier(tx, txErr)
+	if !slices.Contains(sendTxSuccessfulCodes, resultCode) {
+		txSender.lggr.Warnw("RPC returned error", "tx", tx, "err", txErr)
+	}
+	return sendTxResult{Err: txErr, ResultCode: resultCode}
+}
+
+func (txSender *TransactionSender[TX, CHAIN_ID, RPC]) reportSendTxAnomalies(tx TX, txResults <-chan sendTxResult) {
+	defer txSender.wg.Done()
+	resultsByCode := sendTxErrors{}
+	// txResults eventually will be closed
+	for txResult := range txResults {
+		resultsByCode[txResult.ResultCode] = append(resultsByCode[txResult.ResultCode], txResult.Err)
+	}
+
+	_, _, criticalErr := aggregateTxResults(resultsByCode)
+	if criticalErr != nil {
+		txSender.lggr.Criticalw("observed invariant violation on SendTransaction", "tx", tx, "resultsByCode", resultsByCode, "err", criticalErr)
+		PromMultiNodeInvariantViolations.WithLabelValues(txSender.chainFamily, txSender.chainID.String(), criticalErr.Error()).Inc()
+	}
+}
+
+type sendTxErrors map[SendTxReturnCode][]error
+
+func aggregateTxResults(resultsByCode sendTxErrors) (returnCode SendTxReturnCode, txResult error, err error) {
```

Comment:
I find this function a little confusing. For one, the resultsByCode is typed as sendTxErrors, but then we are searching for sendTxSuccessfulCodes down below. Should the type really be something like sendTxCodes?

Next, on line 187/188, it looks like we are searching for successCode and SevereError codes in the results. Is the assumption that if we get a success, that we should never also see a SevereError? I'd typically see this as something like:

if there's a SevereError, return error, else return success. It looks like we need to identify cases where we get both a success and an error?

Reply:
Agree that we can improve the naming of the `sendTxErrors.` I'm not sure about `sendTxCodes` as it's `map[SendTxReturnCode][]error` - results of transaction submission grouped by code. Maybe `sendTxResults`?

We can observe success and severe error for a single transaction in many cases. Here are a couple of examples:
* RPC is misconfigured and rejects tx due to high/low gas while the rest of the network accepts it.
* RPC provider introduced a new error msg for transient error, but we failed to classify it.

In case of a severe error, TXM stops processing new transactions or even abandons current transaction and tries to reuse its nonce. 
We prefer returning success as it's better for Txm to continue processing pending transactions and get stuck if there is actual issue. 
If there is a contradiction in RPCs' responses (success and severe error), we log a critical error to notify NOPs that this issue requires attention.

Reply:
I think `sendTxResults` makes sense, I'll update the naming for now.